[DX] Typed Web Workers

1337LutZ opened the pull request:
Both WASM and Local work great now that you have live changes when running `WATCH=1 make webworkers`

- Converted the workers to Typescript
- Added types to the best of my ability
- Reused shared logic
- Added separate vite build bundling task for compiling each worker in a separate "library
- Added `make webworkers` task to compile, copy & watch the workers

kayla-glick commented:
After running `make wasm` and changing
<img width="400" alt="image" src="https://github.com/wowsims/cata/assets/12898988/8b56ebd8-7133-48e8-b60e-6131b4ee4a6c">
the sim fails to start and throws these errors
<img width="1728" alt="image" src="https://github.com/wowsims/cata/assets/12898988/39bb6c97-d073-4397-b318-597a7ef33d4d">

1337LutZ commented:
> After running `make wasm` and changing <img alt="image" width="400" src="https://private-user-images.githubusercontent.com/12898988/330564490-8b56ebd8-7133-48e8-b60e-6131b4ee4a6c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NTgzNDgsIm5iZiI6MTcxNTg1ODA0OCwicGF0aCI6Ii8xMjg5ODk4OC8zMzA1NjQ0OTAtOGI1NmViZDgtNzEzMy00OGU4LWI2MGUtNjEzMWI0ZWU0YTZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE2VDExMTQwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTViZmIwZTdmNWQzNzU4ZTExYjc4N2ZiOGRiMWZjYjhjMGE3NWVhMjhkY2MxNWFlNThmYTk1ZTRjYjA4NmRhNDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.iwVJpJUbeRoeDGBLrpsLMg328yNb_XhrHHyNsj2hqMQ"> the sim fails to start and throws these errors <img alt="image" width="1728" src="https://private-user-images.githubusercontent.com/12898988/330564437-39bb6c97-d073-4397-b318-597a7ef33d4d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NTgzNDgsIm5iZiI6MTcxNTg1ODA0OCwicGF0aCI6Ii8xMjg5ODk4OC8zMzA1NjQ0MzctMzliYjZjOTctZDA3My00Mzk3LWIzMTgtNTk3YTdlZjMzZDRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MTYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTE2VDExMTQwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ1MTAxMDAyN2UzMGM5ZGQ0NWE1NzkwYTE1ZDVjMTU1MjUyYmI3NzNhMzhiZDgxZjcwODIyNDFiYjMzYmI5YzgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.2l2GPMonJNjV6hvYWmXGTHTyRhUFgvI5IQDpHPmnF84">

Fixed it. `wasmready` was being tree shaken since it was unused. Added it to `globalThis` scope and now it works.

FelixPflaum commented:
I think if we're doing this, then why not refactor them entirely while we're at it. I'm especially offended by the net_ and local_ version sharing like 99.9% of their code :D

Made a quick PR for that against this branch https://github.com/wowsims/cata/pull/482

Advantages:
- All UI interface logic is contained in worker_interface.ts. That also means that adding or removing SimRequest defs in types.ts forces you to add the respective handlers for both worker versions.
- All wasm related code is entirely in sim_worker.ts
- All http related code is in worker_http.ts, the 2 versions then just set a different base URL.

1337LutZ commented:
Awesome, thanks! I will look into merging this 👍

FelixPflaum commented on ui/worker/worker_interface.ts:
```diff
@@ -0,0 +1,62 @@
+
+	/** Tell UI that the worker is ready. */
+	ready() {
+		if (!this.handlers) {
+			console.error('WorkerInterface.ready() used but handlers not set!');
+			return;
+		}
```
Forgot to remove this check. With handlers being set in the constructor it's not needed anymore.

FelixPflaum commented:
For all I care this can be merged. Seems to work fine.

1337LutZ commented:
@rosenrusinov if you validated that the local/sim workers work feel free to merg on your next round of checks :)
